server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java

```diff
@@ -19,48 +19,313 @@
+            return new Id(id);
+        }
+
+        private final String value;
+
+        private Id(String value) {
+            this.value = value;
+        }
+
+        public String asString() {
+            return value;
+        }
+
+        @Override
+        public final boolean equals(Object o) {
+            if (o instanceof Id) {
+                Id id = (Id) o;
+                return Objects.equals(value, id.value);
+            }
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(value);
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                .add("value", value)
+                .toString();
+        }
+    }
+
+    public static class Condition {
+
+        public enum Field {
+            FROM("from"),
+            TO("to"),
+            CC("cc"),
+            SUBJECT("subject"),
+            RECIPIENT("recipient");
+            
+            public static Optional<Field> find(String fieldName) {
+                return Arrays.stream(values())
+                        .filter(value -> value.fieldName.equalsIgnoreCase(fieldName))
+                        .findAny();
+            }
+            
+            public static Field of(String fieldName) {
+                return find(fieldName).orElseThrow(() -> new IllegalStateException("'" + fieldName + "' is not a valid field name"));
+            }
+            
+            private final String fieldName;
+            
+            private Field(String fieldName) {
+                this.fieldName = fieldName;
+            }
+            
+            public String asString() {
+                return fieldName;
+            }
+        }
+        
+        public enum Comparator {
+            CONTAINS("contains"),
+            NOT_CONTAINS("not-contains"),
+            EXACTLY_EQUALS("exactly-equals"),
+            NOT_EXACTLY_EQUALS("not-exactly-equals");
+            
+            public static Optional<Comparator> find(String comparatorName) {
+                return Arrays.stream(values())
+                        .filter(value -> value.comparatorName.equalsIgnoreCase(comparatorName))
+                        .findAny();
+            }
+            
+            public static Comparator of(String comparatorName) {
+                return find(comparatorName).orElseThrow(() -> new IllegalStateException("'" + comparatorName + "' is not a valid comparator name"));
+            }
+            
+            private final String comparatorName;
+            
+            private Comparator(String comparator) {
+                this.comparatorName = comparator;
+            }
+            
+            public String asString() {
+                return comparatorName;
+            }
+        }
+
+        public static Condition of(String field, String comparator, String value) {
```

Comment:
If it's for human being to call, having 3 strings as argument looks like a bad idea

Reply:
I've hesitate to do a `Rule.Condition.Builder`. Should I do it?

Reply:
In this case, I would take Field, Comparator and String as input type.